op-batcher/batcher/driver_test.go

```diff
@@ -160,3 +164,111 @@ func TestBatchSubmitter_sendTx_FloorDataGas(t *testing.T) {
 	expectedFloorDataGas := uint64(21_000 + 12*10)
 	require.GreaterOrEqual(t, candidateOut.GasLimit, expectedFloorDataGas)
 }
+
+// createHTTPHandler creates a mock HTTP handler for testing, it accepts a callback which
+// is invoked when the expected request is received.
+func createHTTPHandler(t *testing.T, cb func(), alwaysFails bool) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Method == "POST" {
+			var req struct {
+				JSONRPC string        `json:"jsonrpc"`
+				Method  string        `json:"method"`
+				Params  []interface{} `json:"params"`
+				ID      interface{}   `json:"id"`
+			}
+			if err := json.NewDecoder(r.Body).Decode(&req); err == nil {
+
+				if alwaysFails {
+					http.Error(w, "Simulated failure", http.StatusInternalServerError)
+					cb()
+					return
+				}
+				if req.Method == "miner_setMaxDASize" && len(req.Params) == 2 {
+					cb()
+					w.Header().Set("Content-Type", "application/json")
+					_, err := w.Write([]byte(`{"jsonrpc":"2.0","id":1,"result":true}`))
+					if err != nil {
+						t.Logf("Error writing response: %v", err)
+					}
+					return
+				}
+			}
+		}
+		http.Error(w, "Unexpected request", http.StatusBadRequest)
+	}
+}
+
+func TestBatchSubmitter_ThrottlingEndpoints(t *testing.T) {
+
+	testThrottlingEndpoints := func(numHealthyServers, numUnhealthyServers int) func(t *testing.T) {
+
+		return func(t *testing.T) {
+			healthyCalls := make([]int, numHealthyServers)
+			unHealthyCalls := make([]int, numUnhealthyServers)
+
+			healthyServers := make([]*httptest.Server, numHealthyServers)
+			unhealthyServers := make([]*httptest.Server, numUnhealthyServers)
+
+			urls := make([]string, 0, numHealthyServers+numUnhealthyServers)
+
+			for i := range healthyCalls {
+				healthyServers[i] = httptest.NewServer(createHTTPHandler(t, func() { healthyCalls[i]++ }, false))
+				urls = append(urls, healthyServers[i].URL)
+				defer healthyServers[i].Close()
+			}
+			for i := range unHealthyCalls {
+				unhealthyServers[i] = httptest.NewServer(createHTTPHandler(t, func() { unHealthyCalls[i]++ }, true))
+				urls = append(urls, unhealthyServers[i].URL)
+				defer unhealthyServers[i].Close()
+			}
+
+			// Setup test context
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			// Add in an endpoint with no server at all
+			urls = append(urls, "http://invalid/")
```

Comment:
This exercises the retry logic when an endpoint goes down for some time (e.g. for maintenance / upgrade).

Reply:
but how does the test establish that an endpoint that is down is eventually brought up again?